### ISSUE TITLE
Support for BCR

### ISSUE THREAD

**Issue by grst** [icbi-lab/scirpy#132]
# BCR support meta issue. 

## Initial PR #183 addresses: 
- [x] change datastructure

 * Instead of `TRA_1`, `TRA_2`, `TRB_1` and `TRB_2` have `arm1_primary`, `arm1_secondary`, `arm2_primary` and `arm2_secondary`
 * have additional 4 columns: `arm1_primary_type`, ...; These can accept values such as `TRA`, `TRB`, `TRG`, `TRD`, `IGH`, `IGL` (see [AIRR locus names](https://docs.airr-community.org/en/latest/datarep/rearrangements.html#locus-names))

 - [x] adapt chain_categories to identify *bona fide* vs other pairings (e.g. flag a cell with TRA + IGH). 
 - [x] clonotype network: separate by receptor_type (per default, no connections between BCR and TCR)
 - [x] update glossary and documentation. Make clear that now there are `VJ` and `VDJ` chains. (mostly done)
 - [x] rename `tcr_` to `cdr3_` or `receptor_` or `vdj_` or `ir_`

## To be resolved before next release (v0.5 "adding experimental BCR support")
 - [x] #194 vdj_usage is broken with new data structure
 - [x] #195 improve BCR-related documentation
 - [x] `read_tracer()` with gamma/delta
 - [x] `read_bracer()` 
 - [x] #198 Add BCR example dataset

## BCR-related issues that can be resolved at a later point
 - [ ] #197 function to infer antibody class
 - [ ] #196 function to infer somatic hypermutation status
 - [ ] #199 BCR-tutorial
 - [ ] add support for CDR1 and 2 (#185)

**Comment by ktpolanski** [comment]
Leaving a line here in case public support can help prioritise development. VDJ analysis is not quite as standardised, it would be great to have a single well-documented package to handle both TCRs and BCRs.

**Comment by grst** [comment]
Hi @ktpolanski,

thanks for your comment! BCR support is definitely THE next thing on the list, atm I'm just a bit overloaded with other tasks. 

Can you, by any chance, recommend a good (publicly available) example dataset with single-cell BCR data that we could use to build the tutorial? 

Cheers, 
Gregor

**Comment by grst** [comment]
@szabogtamas, @FFinotello, @ktpolanski 

Do you guys have any suggestions on how to name the columns for BCR support? 

At the moment we have: 
 * TRA_1_<attr>/TRA_2_<attr>: columns related to the primary and secondary TCR-alpha chain
 * TRB_1_<attr>/TRB_2_<attr>: columns related to the primary and secondary TCR-beta chain
where `<attr>` is `c_gene`, `v_gene`, `cdr3`, `cdr3_nt`, etc. 

This obviously doesn't make sense anymore. 

I thought of 
 * `receptor_arm_1_primary_<attr>`, `receptor_arm_1_secondary_<attr>`, `receptor_arm_2_primary_<attr>` ...
 * `arm1p_<attr>`, `arm1s_<attr>`...

The former is the most explicit, but it's a bit long. Maybe you have better ideas altogether...

**Comment by crushseven-7** [comment]
> @szabogtamas, @FFinotello, @ktpolanski
> 
> Do you guys have any suggestions on how to name the columns for BCR support?
> 
> At the moment we have:
> 
> * TRA_1_/TRA_2_: columns related to the primary and secondary TCR-alpha chain
> * TRB_1_/TRB_2_: columns related to the primary and secondary TCR-beta chain
>   where `<attr>` is `c_gene`, `v_gene`, `cdr3`, `cdr3_nt`, etc.
> 
> This obviously doesn't make sense anymore.
> 
> I thought of
> 
> * `receptor_arm_1_primary_<attr>`, `receptor_arm_1_secondary_<attr>`, `receptor_arm_2_primary_<attr>` ...
> * `arm1p_<attr>`, `arm1s_<attr>`...
> 
> The former is the most explicit, but it's a bit long. Maybe you have better ideas altogether...

About naming header, in my opinion original headline is better than `receptor_arm`. Add a `chain` column might be better to distinguish whether it is IG or TR .

**Comment by massonix** [comment]
Dear scirpy team,

Thanks a lot for developing this amazing tool. I am also highly interested in a tutorial for scBCR-seq. Perhaps one good dataset to illustrate its usage could be found [here](https://www.ebi.ac.uk/arrayexpress/experiments/E-MTAB-9003/). It comes from [this paper](https://www.biorxiv.org/content/10.1101/2020.04.28.054775v2.full.pdf) by Hamish W. King , et al. in which they analyze multiple B cell subsets from palatine tonsils.

Ramon

**Comment by massonix** [comment]
In fact, it'd be awesome if scirpy could infer 3 properties from BCRs that are studied in the paper I shared in the previous comment:

- Clonotype
- Antibody class
- Somatic hypermutation status

I guess clonotype can be assigned in a similar manner to TCR-seq. Antibody class is easily facilitated by the "c_gene" column in "filtered_contig_annotations.csv". Finally, somatic hypermutation might be the most tricky, but maybe you can use a similar approach to the paper: "CreateGermlines.py (ChangeO v0.4.5) was then used to infer germline sequences for each clonal family and observedMutations (Shazam v0.1.11) was used to calculate somatic hypermutation frequencies for each IgH sequence. Sequences with mutation frequencies greater than 0.02 were annotated as “High” mutation levels, those between 0 and 0.02 as “Low” mutation levels and 0 as “None”.

Thanks in advance for considering it!

R

**Comment by grst** [comment]
Hi Ramon, 

thanks for your comments and for sharing that preprint. This is really helpful! 
I'll be on holiday the next weeks, but will look into that end of September! 

Best, 
Gregor

**Comment by massonix** [comment]
Hi Gregor,

It's my pleasure. Thanks for looking into it, I'm highly interested in analyzing scBCR-data using scirpy. I would be more than happy to beta test your software once you have preliminary data.

Best

Ramon

**Comment by grst** [comment]
Just had a short discussion with @FFinotello on column names. 

We currently propose the following naming scheme:

**Previous, only TCR alpha/beta support.**
 * `TRA_1`
 * `TRB_1`
 * `TRA_2`
 * `TRB_2`

**New, with BCR and gamma/delta T-cell support**
 * `IR_CHAIN1_PRI`
 * `IR_CHAIN1_PRI_locus`
 * `IR_CHAIN2_PRI`
 * `IR_CHAIN2_PRI_locus`
 * `IR_CHAIN1_SEC`
 * `IR_CHAIN1_SEC_locus`
 * `IR_CHAIN2_SEC`
 * `IR_CHAIN2_SEC_locus`

where 
 * `CHAIN1` and `CHAIN2` stand for the first and second chain of a receptor pair.
 * `PRI` and `SEC` stand for the primary and secondary receptor of a cell
 * `locus` contains the chain type, e.g. `TRA`, `TRG` or `IGH`.

**Comment by grst** [comment]
After some more thinking, we came up with even more intuitive column names: 

**For the primary receptor:**
 * `IR_VJ_1` (receptor arm with the V-J junction, i.e. TRA, TRD, IGL, IGK)
 * `IR_VDJ_1` (receptor arm with the V-D-J junction, i.e. TRB, TRG, IGH)

**For the secondary receptor**: 
 * `IR_VJ_2`
 * `IR_VDJ_2`

**Comment by grst** [comment]
Basic TCR support has now been added to the master branch. 

I adapted the datastructure to support BCR-data and it should now possible to perform a clonotype-analysis with both B and T-cells. Most parts of the documentation have been updated to reflect the changes. 

Things that don't work yet (see also top of this issue): 
 - [ ] #194 vdj_usage is broken with new data structure
 - [ ] #197 function to infer antibody class
 - [ ] #196 function to infer somatic hypermutation status

I also will be adding a new example dataset (#198). 

@massonix, or anybody else interested: It would be cool if you could give it a try! 
You can install the latest development version of scirpy using
```bash
pip install git+https://github.com/icbi-lab/scirpy.git@master
```

**Comment by grst** [comment]
Just released v0.5.0 which adds support for BCR and gamma/delta TCR :tada:.

**Comment by grst** [comment]
Just discovered https://github.com/zktuong/dandelion, which wraps tools from the [immcantation](https://bitbucket.org/kleinstein/immcantation/src/master/) suite in a scanpy/scirpy-like python interface. 

This could be useful for addressing the remaining BCR issues.

**Comment by zktuong** [comment]
@grst Hi, Yes sure. I'm just in the process of updating the notebooks and documentation to make it compatible with readthedocs. So hopefully once that is done we can discuss a bit further. The tool should be accompanying a manuscript that would probably be up on a pre-print server soon.

**Comment by grst** [comment]
great, looking forward to reading it!

**Comment by zktuong** [comment]
OK the docs are now live at [https://sc-dandelion.readthedocs.io/](https://sc-dandelion.readthedocs.io/en/master/).

**Comment by zktuong** [comment]
Hi @grst, 

I finally have some time to route back to this; sorry for the slow/delayed response, and sorry for the long post below.

I've been reworking the data structure for `dandelion`'s class object to remove some esoteric code and facilitate a future possibility of TCR analysis and also integration with other tools such as `scripy`. It's not quite there yet but the major hurdle should now be removed. Previously, `dandelion` will fail to initialize properly if a cell have multiple IGH contigs but this is now solved. 

Anyway, rather than reinventing the wheel, for now I think it makes more sense to be able to retrieve the data that is processed/analyzed via `scirpy` from the `AnnData` object, to access the other functions that requires the AIRR format that `dandelion` is primarily working with. I also like the options that you currently have for TCR clone definitions but I think I still want to write a function for `dandelion` to try and name the clone_ids to reflect the V-J gene usages, CDR3 length and alpha chain pairings.

There's obviously a few differences in how the data tables are formatted between `dandelion` and `scirpy` but I think it mostly boils down to column dame differences. `dandelion` (at least from v0.0.28 onwards) can be flexible in terms of multiple-contigs mapping to a single cell, for example:

The default schema that i'm using to construct a more readable data structure for now is this:
(A) assuming multiple heavy chain and light chain contigs:
`v_call_heavy, v_call_light`
`IGHV3-1|IGHV3-2|IGHV3-3, IGLV1-2|IGLV2-1|IGKV2-2`

The user can then retrieve the data separately if splitting the following ways:
(B) if split based on whether it's heavy chain/light chain
`v_call_heavy_0, v_call_heavy_1, v_call_heavy_2, v_call_light_0, v_call_light_1, v_call_light_2 ...`
`IGHV3-1, IGHV3-2, IGHV3-3, IGLV1-2, IGLV2-1, IGKV2-2`
(C) if split by locus:
`v_call_IGH_0, v_call_IGH_1, v_call_IGH_2, v_call_IGK_0, v_call_IGK_1, v_call_IGL_0, v_call_IGL_1 ...`
`IGHV3-1, IGHV3-2, IGHV3-3, IGLV1-2, IGLV2-1, IGKV2-2`

By default, numerical columns such as `umi counts` are always splitted according to (B) so as not to create issues with multiple dtypes occuring in the table.

In cellranger v4 onwards, there is an AIRR schema that comes off automatically so IMO that's probably the easiest place to start to try and test some form of integration. It's still lacking a few columns that `dandelion` needs to work fully but most of it is solveable. I wonder if you are already working on read function for the AIRR schema to read it directly to `scirpy`? 

Pre-cellranger V4, a simple dicionary maybe a way to try and sync up the two objects? 

What do you think?

**Comment by grst** [comment]
Hi Kelvin, 

thanks for the initiative. I would absolutely love if we managed to settle on a common datastructure.

I think the challenge with AIRR data is that AIRR data is chain-centric, while AnnData is cell-centric. 
Since all the scanpy functions operate on cells, I finally chose to squeeze the receptor data into a cell-centric
format, by adding each column from the AIRR rearrangement schema four times, supporting both a primary and secondary VJ and VDJ chain, discarding excessive chains by flagging the cells as multichain. It is a pragmatic solution, but I'm not sure it's the optimal solution. I'm not sure if I fully understood whether you opted for a cell-centric of chain centric approach in Dandelion? 

FWIW, all input function in scirpy first go through a chain-centric format: Each cell is an [IrCell](https://icbi-lab.github.io/scirpy/generated/scirpy.io.IrCell.html#scirpy.io.IrCell) which can have one or multiple [IrChains](https://icbi-lab.github.io/scirpy/generated/scirpy.io.IrChain.html#scirpy.io.IrChain) (See also the [data loading tutorial](https://icbi-lab.github.io/scirpy/tutorials/tutorial_io.html#Creating-AnnData-objects-from-other-formats)). This could actually be a starting point for a common format: 
Rewrite those classes that they are fully AIRR compliant, and use them both in scirpy and dandelion as intermediate format that can be used by both tools, and exported back to the AIRR rearrangement schema. 

Another route would be to push https://github.com/theislab/anndata/issues/237 into a direction that makes sense to both of us. 

> In cellranger v4 onwards, there is an AIRR schema that comes off automatically so IMO that's probably the easiest place to start to try and test some form of integration. It's still lacking a few columns that dandelion needs to work fully but most of it is solveable. I wonder if you are already working on read function for the AIRR schema to read it directly to scirpy?

If you are talking about the AIRR rearrangement schema, there's already [scirpy.io.read_airr](https://icbi-lab.github.io/scirpy/generated/scirpy.io.read_airr.html#scirpy.io.read_airr). I currently only pick the columns that I need, and the columns in anndata are named slightly different than in AIRR, mostly because I didn't know about it when I initially drafted the format. I'd be happy to change that, though.

**Comment by zktuong** [comment]
Hmm I would say dandelion is primarily chain-centric and input works very similarly to your description of scirpy's but not as fancy. 

I basically read the entire AIRR table and wrote a parser to interact with the AIRR table directly with a few standard output columns as part of the initialization procedure. All other functions work off the original AIRR format and then returns it directly to the AIRR table, which i can then loop the parser to convert it to be `AnnData.obs` friendly.

I think it would be relatively easy for me to write dandelion -> scirpy converter based on your `scirpy.io.read_airr` function (because dandelion's core data slot is essentially just the entire AIRR table) but wondering about the reverse. Right now, I dandelion primarily needs the locus, v_call, j_call, c_call, cell_id, sequence_id, junction_aa (or junction), sequence_alignment, germline_alignment to a initialize. So perhaps the reverse would be to make IrCell-like class to be able to extract and hold all these information from a scirpy-processed h5ad and then I can parse that out? Other things like distances and connectivities should be simple to the copy and paste into dandelion<->scirpy directly via AnnData.

**Comment by grst** [comment]
Great! 
There's already a [scirpy.io.to_ir_objs](https://icbi-lab.github.io/scirpy/generated/scirpy.io.to_ir_objs.html#scirpy.io.to_ir_objs) function that can convert a scirpy anndata back to a list of `IrCells`. Should be straightforward from there. Actually, I could also consider implementing a `write_airr` function. 

Just note that excessive chains will be lost in that process - I don't know yet how you deal with them.

**Comment by zktuong** [comment]
Sounds good. Just want to flag a couple of hiccups:

1) `scirpy.io.to_ir_objs` doesn't work on the adata downloaded using `adata = ir.datasets.wu2020_3k()`
the following error is raised:
`ValueError: Key 'IR_VJ_1_expr_raw' not found in adata. Does it contain immune receptor data?`

2) `scirpy.io.read_airr` doesn't work on the 10x cellranger-generated `airr_rearrangement.tsv` because the file annoyingly lacks the `locus` column. Could probably infer from the v-, j-, c- calls to check if they contains TR[ABDG]/IG[HKL] otherwise flag as failed contig?

**Comment by grst** [comment]
Thanks for reporting, I'll try to address them this week.

**Comment by zktuong** [comment]
I've created a draft PR #240 and added the code that deals wtih (2) and (3) above but not annotated fully (because i don't know how to haha).

Also I think what can be improved is if the `IrCell` can also stash `clone_id` if there is that is annotated in some form in the h5ad/airr table. Even better would be if `IrCell` can be interated so that it will just store whatever information it sees that is IR related.

**Comment by grst** [comment]
Cool!

> Even better would be if IrCell can be interated so that it will just
store whatever information it sees that is IR related.

Yes, let's make it an AirrCell that's fully AIRR compliant.

**Comment by grst** [comment]
I thought a bit more about how to restructure the scirpy IO module: 

 * Rename `IrCell` and `IrChain` to `AirrCell` and `AirrChain` to reflect that they are fully AIRR compliant
 * Change them to support all columns defined in the rearrangement schema. Except the fields required by the AIRR standard all fields are optional. 
 * I found a way (although a hacky one: pickling and base64 encoding) of storing excessive chains in an extra chain. They would not be used by scirpy, but it would make reading AIRR files non-destructive. 
 * Implement a `write_airr` function in addition to the `{to,from}_dandelion` functions you are working on already. 

I think this would make scirpy a lot more interoperable, and allow to use the IO module to convert legacy 10x and Tracer/Bracer to AIRR format. 

How does that sound? Have I missed anything?

**Comment by zktuong** [comment]
I think that all sounds good. It would be great if the extra chains information can be stored somewhere and retrieved appropriately as you described.

I also just noticed a minor parsing 'failure' when converting from scirpy -> dandelion but not sure if it's going to be an issue but i can forsee some potential confusion. These are cells without IGH contigs (which I presumed would be removed from scirpy's processing later on anyway) and they would be present in their contig/chain form in both scirpy and dandelion but not parsed as cells in dandelion (but yes in scirpy). Anyway just to note this here.

**Comment by grst** [comment]
I am closing this as scirpy now includes an interface to [Dandelion](https://github.com/zktuong/dandelion), which provides 
convenient wrappers about [Immcantation tools](https://immcantation.readthedocs.io/en/stable/) which should 
cover most of the use-cases discussed above. 

If anyone has more specific requests/ideas regarding BCR analysis, feel free to open a separate issue.
